search-provider/yelp-search-provider.c

```diff
@@ -0,0 +1,587 @@
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program; if not, see <http://www.gnu.org/licenses/>.
+ *
+ * Author: Lili Cosic <lili@kinvolk.io>
+ */
+
+#include <config.h>
+
+#include <gio/gio.h>
+#include <string.h>
+
+#include "yelp-shell-search-provider-generated.h"
+
+#include "libyelp/yelp-document.h"
+#include "libyelp/yelp-uri.h"
+
+#define SEARCH_PROVIDER_INACTIVITY_TIMEOUT 12000 /* milliseconds */
+
+struct _PageData
+{
+    gchar *title;
+    gchar *title_casefold;
+    gchar *desc;
+    gchar *desc_casefold;
+    GIcon *icon;
+};
+
+typedef struct _PageData PageData;
+
+typedef GApplicationClass YelpSearchProviderAppClass;
+typedef struct _YelpSearchProviderApp YelpSearchProviderApp;
+
+struct _YelpSearchProviderApp {
+    GApplication parent;
+    gulong uri_resolve_id;
+    YelpShellSearchProvider2 *skeleton;
+
+    GHashTable *page_data_hash_map;
+    GPtrArray *delayed_result_getters;
+};
+
+GType yelp_search_provider_app_get_type (void);
+
+#define YELP_TYPE_SEARCH_PROVIDER_APP yelp_search_provider_app_get_type()
+#define YELP_SEARCH_PROVIDER_APP(obj) \
+      (G_TYPE_CHECK_INSTANCE_CAST((obj), YELP_TYPE_SEARCH_PROVIDER_APP, YelpSearchProviderApp))
+
+G_DEFINE_TYPE (YelpSearchProviderApp, yelp_search_provider_app, G_TYPE_APPLICATION)
+
+static PageData *
+page_data_new_steal (gchar *title,
+                     gchar *desc,
+                     GIcon *icon)
+{
+    PageData *data = g_new0 (PageData, 1);
+    if (title) {
+        data->title = title;
+        data->title_casefold = g_utf8_casefold (title, -1);
+    }
+    if (desc) {
+        data->desc = desc;
+        data->desc_casefold = g_utf8_casefold (desc, -1);
+    }
+    data->icon = icon;
+    return data;
+}
+
+static void
+page_data_free (PageData *page_data)
+{
+    g_free (page_data->title);
+    g_free (page_data->title_casefold);
+    g_free (page_data->desc);
+    g_free (page_data->desc_casefold);
+    g_clear_object (&page_data->icon);
+    g_free (page_data);
+}
+
+static GVariant *
+get_result_metas (gchar      **page_ids,
+                  GHashTable  *page_data)
+{
+    GVariantBuilder metas;
+    gint i;
+
+    g_variant_builder_init (&metas, G_VARIANT_TYPE ("aa{sv}"));
+
+    for (i = 0; page_ids[i] != NULL; i++) {
+        GVariantBuilder meta;
+        PageData *data = g_hash_table_lookup (page_data, page_ids[i]);
+        g_variant_builder_init (&meta, G_VARIANT_TYPE ("a{sv}"));
+        g_variant_builder_add (&meta, "{sv}",
+                              "id", g_variant_new_string (page_ids[i]));
+        if (data != NULL) {
```

Comment:
I don't think there's any verification (or obligation from the provider) that a meta is provided for each passed-in result, so I believe the `data == NULL`  case could be removed